lib/IRGen/GenStruct.cpp

```diff
@@ -1517,11 +1518,15 @@ class ClangRecordLowering {
     unsigned explosionEnd = NextExplosionIndex;
 
     ElementLayout layout = ElementLayout::getIncomplete(fieldType);
-    auto isEmpty = fieldType.isKnownEmpty(ResilienceExpansion::Maximal);
-    if (isEmpty)
-      layout.completeEmptyTailAllocatedCType(
-          fieldType.isTriviallyDestroyable(ResilienceExpansion::Maximal), NextOffset);
-    else
+    auto isEmpty = isZeroSized || fieldType.isKnownEmpty(ResilienceExpansion::Maximal);
+    if (isEmpty) {
+      if (isZeroSized)
+        layout.completeEmpty(
```

Comment:
Don't have a lot of context about the interop side of things, but just wanted to point out that isZeroSized does not imply empty type (i.e., `[[no_unique_address]]` on non-empty field allows Clang to fold other fields into the tail padding of that non-empty NUA field).

Reply:
https://github.com/llvm/llvm-project/issues/50766

Reply:
Ah looks like Clang still doesn't currently make use of that (but it's a thing that could start happening, https://github.com/llvm/llvm-project/pull/90462)

Reply:
Thanks, that is a good point. Once Clang starts to reuse the tail padding we definitely should have some tests for that. I do remember us running into some trouble when a base class' tail padding was reused, Egor did some fixes in that space. 

Here, `isZeroSized` comes from `FieldDecl::isZeroSize` and it is documented as `"Determine if this field is a subobject of zero size, that is, either a zero-length bit-field or a field of empty class type with the [[no_unique_address]] attribute."`. I think based on this, it should be false for non-empty fields.

Reply:
One quirk is for example:
```
struct E1 {};
struct E2 { [[no_unique_address]] E1 e; };
struct Foo { [[no_unique_address]] E2 e; };
```
where `isZeroSize` on `Foo::e` would also be `true`. Just brought it up in case it brings some edge-cases to mind :)

> Once Clang starts to reuse the tail padding we definitely should have some tests for that.
Sounds good! 👍